models/pattern/stages/import.go

```diff
@@ -0,0 +1,280 @@
+Otherwise we push the service onto importingstack. This happens until the importingstack is empty.
+*/
+type servicewrapper struct {
+	parentname string
+	svc        *core.Service
+	name       string
+	imported   map[string]bool
+}
+type servicestack struct {
+	sws []*servicewrapper
+}
+
+func (ss *servicestack) push(s *servicewrapper) {
+	ss.sws = append(ss.sws, s)
+}
+func (ss *servicestack) pop() (s *servicewrapper) {
+	if ss.isEmpty() {
+		return nil
+	}
+	svc := ss.sws[len(ss.sws)-1]
+	ss.sws = ss.sws[0 : len(ss.sws)-1]
+	return svc
+}
+func (ss *servicestack) isEmpty() bool {
+	if len(ss.sws) == 0 {
+		return true
+	}
+	return false
+}
+func (ss *servicestack) runOnStack(fn func(svc *core.Service)) {
+	for _, s := range ss.sws {
+		fn(s.svc)
+	}
+}
+func Import(prov ServiceInfoProvider, act ServiceActionProvider) ChainStageFunction {
+	return func(data *Data, err error, next ChainStageNextFunction) {
+		if err != nil {
+			act.Terminate(err)
+			return
+		}
+		data.Lock.Lock()
+		//If no imports are there, completely bypass this stage
+		for _, svc := range data.Pattern.Services {
+			if _, ok := matchImportPattern(svc.Type); ok { //If we found any service with imports
+				nonImportingServiceStack := &servicestack{sws: nil}
+				importingServiceStack := &servicestack{}
+				vars := map[string]interface{}{}
+				for _, svc := range data.Pattern.Services {
+					importingServiceStack.push(&servicewrapper{
+						parentname: "",
+						svc:        svc,
+						name:       svc.Name,
+						imported:   map[string]bool{},
+					})
+				}
+				//At the end of this processing, importingServiceStack will be empty and nonImportingServiceStack will have all the services in it with proper dependson and references set
+				err = process(importingServiceStack, nonImportingServiceStack, vars)
+				if err != nil {
+					act.Terminate(err)
+					return
+				}
+				data.Pattern.Services = stackToServices(nonImportingServiceStack)
+				data.Pattern.Vars = vars
+				patternYaml, err := yaml.Marshal(data.Pattern)
+				if err != nil {
+					if err != nil {
+						act.Terminate(err)
+						return
+					}
+				}
+				*data.Pattern, err = core.NewPatternFile(patternYaml)
+				if err != nil {
+					act.Terminate(err)
+					return
+				}
+			}
+			break
+		}
+		data.Lock.Unlock()
+		if next != nil {
+			next(data, nil)
+		}
+	}
+}
+
+func stackToServices(s *servicestack) (services map[string]*core.Service) {
+	services = make(map[string]*core.Service)
+	if s.isEmpty() {
+		return
+	}
+	for _, svcwrap := range s.sws {
+		services[svcwrap.name] = svcwrap.svc
+	}
+	return
+}
+func process(imp *servicestack, nonimp *servicestack, vars map[string]interface{}) error {
+	for !imp.isEmpty() {
+		sw := imp.pop()
+		url, ok := matchImportPattern(sw.svc.Type)
+		if !ok {
+			nonimp.push(sw)
+			continue
+		}
+		p, err := getPatternFromLocation(url)
```

Comment:
wouldn't we want to validate this pattern file before we start processing it 🤔 ? @Revolyssup